src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -361,7 +361,10 @@ internal override ValueTask ConnectAsync(CancellationToken cancellationToken = d
             QUIC_CONNECTION_SHUTDOWN_FLAGS Flags,
             long ErrorCode)
         {
-            Debug.Assert(!_state.ShutdownTcs.Task.IsCompleted);
+            if (_state.ShutdownTcs.Task.IsCompleted)
+            {
+                return ValueTask.CompletedTask;
```

Comment:
If we want to have a fair no-op, we should check that flags are the same as they were the last time... or at least check `IsCompletedSuccessfully` instead of `IsCompleted` - otherwise we would not be able to send an abortive shutdown after we timeouted waiting for graceful shutdown to complete (though, I'm not 100% sure msquic supports it 😬)

Reply:
You're right, this is very shortsighted change. I'll just leave out the assert and handle the multiple calls to `ShutdownAsync` on the `Http3Connection` side.